include/deal.II/rol/vector_adaptor.h

```diff
@@ -0,0 +1,478 @@
+   * VectorType::begin();
+   *
+   *                // Iterator to the one past the last (locally owned)
+   *                // element of the current vector.
+   * VectorType::end();
+   *
+   *                     // Compress the vector i.e., flush the buffers of the
+   *                     // vector object if it has any.
+   * VectorType::compress(VectorOperation::insert);
+   * @endcode
+   *
+   * Most of the vectors in deal.II (see Vector classes) adhere to the above
+   * requirements.
+   *
+   * The current implementation in ROL doesn't support vector sizes above
+   * the largest value of int type.
+   */
+  template<typename VectorType>
+  class VectorAdaptor : public ROL::Vector<typename VectorType::value_type>
+  {
+
+    /**
+     * A typedef for size type of VectorType.
+     */
+    using size_type = typename VectorType::size_type;
+
+    /**
+     * A typedef for element type stored in the VectorType.
+     */
+    using value_type = typename VectorType::value_type;
+
+    /**
+     * A typedef for real-valued numbers.
+     */
+    using real_type = typename VectorType::real_type;
+
+  private:
+
+    /**
+     * Teuchos smart reference counting pointer to the Vector.
+     */
+    Teuchos::RCP<VectorType> vector_ptr;
+
+  public:
+
+    /**
+     * Constructor.
+     */
+    VectorAdaptor (const Teuchos::RCP<VectorType> &vector_ptr);
+
+    /**
+     * Returns the Teuchos smart reference counting pointer to the Vector,
+     * #vector_ptr.
+     */
+    Teuchos::RCP<VectorType> getVector ();
+
+    /**
+     * Returns the Teuchos smart reference counting pointer to const Vector.
+     */
+    Teuchos::RCP<const VectorType> getVector () const;
+
+    /**
+     * Return the dimension of the Vector.
+     */
+    int dimension () const;
+
+    /**
+     * Set the Vector to a given ROL::Vector @p rol_vector by overwriting its
+     * contents.
+     *
+     * If the current Vector has ghost elements,
+     * then VectorType::operator=(const VectorType&) should still be allowed
+     * on the current Vector.
+     */
+    void set (const ROL::Vector<value_type> &rol_vector);
+
+    /**
+     * Perform addition.
+     */
+    void plus (const ROL::Vector<value_type> &rol_vector);
+
+    /**
+     * Scale the Vector by @p alpha and add ROL::Vector @p rol_vector to it.
+     */
+    void axpy (const value_type               alpha,
+               const ROL::Vector<value_type> &rol_vector);
+
+    /**
+     * Scale the Vector.
+     */
+    void scale (const value_type alpha);
+
+    /**
+     * Return the dot product with a given ROL::Vector @p rol_vector.
+     */
+    value_type dot (const ROL::Vector<value_type> &rol_vector) const;
+
+    /**
+     * Return the \f$ L_2 \f$ norm of the Vector.
+     *
+     * The returned type is of VectorAdaptor::value_type so as to maintain
+     * consistency with ROL::Vector<VectorAdaptor::value_type> and
+     * more importantly to not to create an overloaded version namely,
+     * @code
```

Comment:
you can do here `<code>VectorAdaptor::real_type norm() const;</code>` so that it's easier to read the sentence.